test/e2e/parallel/serviceaccount_test.go

```diff
@@ -36,35 +35,64 @@ func TestDoNotOverrideServiceAccount(t *testing.T) {
 		Resources()
 
 	// and move the user to appstudio tier
-	tiers.MoveSpaceToTier(t, awaitilities.Host(), mur.Name, "appstudio")
+	tiers.MoveSpaceToTier(t, awaitilities.Host(), mur.Name, "appstudio-env")
 	VerifyResourcesProvisionedForSpace(t, awaitilities, mur.Name)
 
-	// get the SA that is provisioned for the user in the ns
-	sa, err := member.WaitForServiceAccount(t, mur.Name, fmt.Sprintf("appstudio-%s", mur.Name))
-	require.NoError(t, err)
-	expectedSecrets := getSASecrets(t, member, mur.Name, sa.Name)
+	expectedSecrets := getSASecrets(t, member, mur.Name, "namespace-manager")
 
-	// when we add an annotation to the SA resource then it should stay there
-	sa.Annotations = map[string]string{
-		"should": "stay",
-	}
-	require.NoError(t, member.Client.Update(context.TODO(), sa))
+	for i := 0; i < 10; i++ {
+		_, err := member.UpdateServiceAccount(t, fmt.Sprintf("%s-env", mur.Name), "namespace-manager", func(sa *corev1.ServiceAccount) {
 
-	// drop the SpaceRoles annotation from the namespace to trigger the reconciliation
-	require.NoError(t, err)
-	_, err = member.UpdateNSTemplateSet(t, mur.Name, func(nsTmplSet *v1alpha1.NSTemplateSet) {
-		delete(nsTmplSet.Annotations, v1alpha1.LastAppliedSpaceRolesAnnotationKey)
-	})
-	require.NoError(t, err)
+			// when we add an annotation to the SA resource then it should stay there
+			sa.Annotations = map[string]string{
+				"should": "stay",
+			}
+			sa.Secrets = append(sa.Secrets, corev1.ObjectReference{
+				Name: fmt.Sprintf("dummy-secret-%d", i),
+			})
+			sa.ImagePullSecrets = append(sa.ImagePullSecrets, corev1.LocalObjectReference{
+				Name: fmt.Sprintf("dummy-pull-secret-%d", i),
+			})
 
-	// then
-	VerifyResourcesProvisionedForSpace(t, awaitilities, mur.Name)
-	sa, err = member.WaitForServiceAccount(t, mur.Name, fmt.Sprintf("appstudio-%s", mur.Name))
-	require.NoError(t, err)
-	assert.Equal(t, "stay", sa.Annotations["should"])
+		})
+		require.NoError(t, err)
+
+		// drop the SpaceRoles annotation from the namespace to trigger the reconciliation
+		_, err = member.UpdateNSTemplateSet(t, mur.Name, func(nsTmplSet *v1alpha1.NSTemplateSet) {
+			delete(nsTmplSet.Annotations, v1alpha1.LastAppliedSpaceRolesAnnotationKey)
+		})
+		require.NoError(t, err)
+
+		// then
+		VerifyResourcesProvisionedForSpace(t, awaitilities, mur.Name)
+		sa, err := member.WaitForServiceAccount(t, fmt.Sprintf("%s-env", mur.Name), "namespace-manager")
+		require.NoError(t, err)
+		assert.Equal(t, "stay", sa.Annotations["should"])
+	secrets:
+		for j := 0; j < i; j++ {
+			expName := fmt.Sprintf("dummy-secret-%d", j)
+			for _, secretRef := range sa.Secrets {
+				if secretRef.Name == fmt.Sprintf("dummy-secret-%d", j) {
+					continue secrets
+				}
+			}
+			assert.Fail(t, fmt.Sprintf("secret '%s' not found", expName))
+		}
```

Comment:
just trying to understand - when `i` is 0 this check will be skipped (ok it will be checked once i>0).
But why do we need to start from `secrets` again every time we find a secret?

Could we check that all the expected number of secrets (j) are there , and fail the the first one that is missing ?  

something like:
```
	for j := 0; j <= i; j++ {
		expName := fmt.Sprintf("dummy-secret-%d", j)
		found := false
		for _, secretRef := range sa.Secrets {
			if secretRef.Name == fmt.Sprintf("dummy-secret-%d", j) {
			        found = true
				break
			}
		}
		if !found {
			assert.Fail(t, fmt.Sprintf("secret '%s' not found", expName))
		}
	}
``` 
and remove the `secrets` tag and continue statement ?

Reply:
> just trying to understand - when i is 0 this check will be skipped (ok it will be checked once i>0).

right - bad copy-paste from the `for i` line :-) 

as for your version - isn't it the same?
the only benefit would be that I could combine secrets & pullSecrets together

Reply:
If I'm not mistaken is not the same because:
1. you can combine them right
2. you are not starting from j=0 every time you find a secret, but you validate them in one single loop, because the current `continue secrets` exists the outer loop as well right ?